### PR TITLE
DataGrid: Fix added rows visibility after scrolling if scrolling mode and rowRenderingMode are "virtual" (T939302)

### PR DESCRIPTION
**Pull request by vconst**

### CODE REVIEW

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 1]
```diff
@@ -799,10 +799,28 @@ const EditingController = modules.ViewController.inherit((function() {
...
             return item;
         },
-
-        processItems: function(items, { changeType, removeCount }) {
+        _getLoadedRowIndexByInsertKey: function(items, change, key) {
             const dataController = this._dataController;
-            let dataRowIndex = -1;
+            const loadedRowIndexOffset = dataController.getRowIndexOffset(true);
+            let loadedRowIndex = key.dataRowIndex - loadedRowIndexOffset;
+
```
Comment (NickMitrokhin) [inline js/ui/grid_core/ui.grid_core.editing.js:806]: loadedRowIndex may be NaN if key does not have the dataRowIndex option. For instance, if the change type is not 'insert'.

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 2]
```diff
@@ -811,19 +829,11 @@ const EditingController = modules.ViewController.inherit((function() {
...
-                            dataRowIndex += removeCount;
-                        }
-                    }
+                    const loadedRowIndex = this._getLoadedRowIndexByInsertKey(items, e, key);
 
                     const item = this._generateNewItem(key);
-                    if(dataRowIndex >= 0 && change.type === DATA_EDIT_DATA_INSERT_TYPE && this._needInsertItem(change, changeType, items, item)) {
-                        items.splice(key.dataRowIndex ? dataRowIndex : 0, 0, item);
+                    if(loadedRowIndex >= 0 && change.type === DATA_EDIT_DATA_INSERT_TYPE && this._needInsertItem(change, changeType, items, item)) {
+                        items.splice(key.dataRowIndex ? loadedRowIndex : 0, 0, item);
```
Comment (NickMitrokhin) [inline js/ui/grid_core/ui.grid_core.editing.js]: From what I gather, all this logic is intended for the 'insert' change type. Isn't it better to change the code a little bit as shown below?
```js
if(key && change.type === DATA_EDIT_DATA_INSERT_TYPE) {
    const loadedRowIndex = this._getLoadedRowIndexByInsertKey(items, e, key);

    const item = this._generateNewItem(key);
    if(loadedRowIndex >= 0 && this._needInsertItem(change, changeType, items, item)) {
            items.splice(key.dataRowIndex ? loadedRowIndex : 0, 0, item);
    }
}
```
In this case _getLoadedRowIndexByInsertKey will not return NaN using the current implementation.

File: js/ui/grid_core/ui.grid_core.virtual_scrolling.js [hunk 3]
```diff
@@ -808,7 +807,7 @@ export default {
 
                         that._visibleItems = [];
 
-                        const isItemCountable = function(item) {
+                        const isItemCountable = function(item, isNextAfterLast, fromEnd) {
```
Comment (NickMitrokhin) [inline js/ui/grid_core/ui.grid_core.virtual_scrolling.js]: why **isNextAfterLast** and **fromEnd** are required here?

File: testing/tests/DevExpress.ui.widgets.dataGrid/dataController.tests.js [hunk 4]
```diff
@@ -3972,6 +3975,130 @@ QUnit.module('Virtual rendering', { beforeEach: setupVirtualRenderingModule, aft
...
+    QUnit.test('addRow > scroll to far > scroll back', function(assert) {
+        // arrange
+        this.addRow();
+
+        // act
+        this.dataController.setViewportPosition(500);
+
+        // assert
+        assert.strictEqual(this.dataController.items()[0].key, 50, 'first item is new');
+
```
Comment (NickMitrokhin) [inline testing/tests/DevExpress.ui.widgets.dataGrid/dataController.tests.js]: Why an item whose key value is 50 is considered as 'new'?

**Review by NickMitrokhin — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets.dataGrid/dataController.tests.js (hunk 4 above)
Comment (vconst) [reply testing/tests/DevExpress.ui.widgets.dataGrid/dataController.tests.js]: t's copy/paste. I will fix it.

**Review by NickMitrokhin — APPROVED** [review]

**Review by Alyar666 — APPROVED** [review]

**Pull request merged**